src/storage/mvcc/txn.rs

```diff
@@ -195,6 +195,61 @@ impl<'a, T: Engine + ?Sized> MvccTxn<'a, T> {
     }
 }
 
+pub struct SnapshotTxn<'a, T: Snapshot + ?Sized + 'a> {
+    snapshot: &'a T,
+    start_ts: u64,
+}
+
+impl<'a, T: Snapshot + ?Sized> fmt::Debug for SnapshotTxn<'a, T> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "snapshot txn @{}", self.start_ts)
+    }
+}
+
+impl<'a, T: Snapshot + ?Sized> SnapshotTxn<'a, T> {
+    pub fn new(snapshot: &'a T, start_ts: u64) -> SnapshotTxn<'a, T> {
+        SnapshotTxn {
+            snapshot: snapshot,
+            start_ts: start_ts,
+        }
+    }
+
+    fn load_meta(&self, key: &Key) -> Result<Meta> {
+        let meta = match try!(self.snapshot.get(key)) {
+            Some(x) => try!(Meta::parse(&x)),
+            None => Meta::new(),
+        };
+        Ok(meta)
+    }
+
+    pub fn get(&self, key: &Key) -> Result<Option<Value>> {
+        let meta = try!(self.load_meta(key));
+        self.get_impl(key, &meta, self.start_ts.to_owned())
+    }
+
+    fn get_impl(&self, key: &Key, meta: &Meta, ts: u64) -> Result<Option<Value>> {
+        // Check for locks that signal concurrent writes.
+        if let Some(lock) = meta.get_lock() {
+            if lock.get_start_ts() <= ts {
+                // There is a pending lock. Client should wait or clean it.
+                return Err(Error::KeyIsLocked {
+                    key: key.raw().to_owned(),
+                    primary: lock.get_primary_key().to_vec(),
+                    ts: lock.get_start_ts(),
+                });
+            }
+        }
+        // Find the latest write below our start timestamp.
+        match meta.iter_items().find(|x| x.get_commit_ts() <= ts) {
```

Comment:
No need. Since timestamps are stored in desc order. We are able to get the version we want for a few iterations.